Migrate CircleCI workflows to GitHub Actions (3/3)

shovnik opened the pull request:
## Which problem is this solving?

As part of [issue](https://github.com/open-telemetry/opentelemetry-collector/issues/1234), this pull request completes the migration to GitHub Actions for the core repo by migrating the remaining `windows-msi` and publish jobs to GHA. The `publish-check` job was intentionally omitted as it seems to only print a success or fail message of the binaries and distributions were successfully created, but does not actually accomplish any task. Let me know if it served an important purpose I might have missed.

When merging this PR, once the repository is ready to move publishing from CircleCI to GHA, a maintainer will need to add the DOCKER_USERNAME and DOCKER_PASSWORD secrets for the opentelemetry docker account to allow pushing to docker. In addition a token with all access writes will need to be generated and added as a GITHUB_TOKEN secret to allow releases.

Important: This PR can only be merged and properly reviewed after part 2/3 has been merged as it branches off that PR: https://github.com/open-telemetry/opentelemetry-collector/pull/2298.

## Migration Plan
We suggest having CircleCI and GitHub Action jobs run in parallel for a few weeks. After the GitHub Actions jobs are running fine for a week or so and then remove the CircleCI workflows from config.yml
cc- @alolita, @AzfaarQureshi

shovnik commented on internal/buildscripts/packaging/msi/make.ps1:
```diff
@@ -65,10 +69,8 @@ function Confirm-MSI {
+    # uninstall msi
     Start-Process -Wait msiexec "/x `"$msipath`" /qn"
-    sc.exe query state=all | findstr "otelcol" | Out-Null
-    if ($LASTEXITCODE -ne 1) { Throw "otelcol service failed to uninstall" }
```
The uninstall validation had to be removed because on GitHub Actions an exit code 1 will immediately fail the workflow. This test attempts to query an uninstalled service to verify it is in fact uninstalled and handles the error.

shovnik commented on internal/buildscripts/packaging/msi/make.ps1:
```diff
@@ -36,6 +36,7 @@ function Install-Tools {
 
     choco install wixtoolset -y
     setx /m PATH "%PATH%;C:\Program Files (x86)\WiX Toolset v3.11\bin"
+    echo "C:\Program Files (x86)\WiX Toolset v3.11\bin" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
```
This line is required to add the WiX Toolset to the path on GitHub Actions. Once CircleCI is removed, the above line: "setx /m ..." will need to be removed.

shovnik commented on .github/workflows/scripts/verify-dist-files-exist.sh:
```diff
@@ -0,0 +1,21 @@
+        exit 0 
+    fi
+done
+echo "::set-output name=passed::true"
```
Extracted repeated shell commands to script that is called instead.

bogdandrutu commented:
this needs a rebase.

shovnik commented:
One sec, new issue that came up after rebase due to changing upload/downlad versions.

shovnik commented:
@bogdandrutu The PR is now rebased and ready for review.

bogdandrutu commented:
More rebase is needed :)

bogdandrutu commented on .github/workflows/build-and-test.yml:
```diff
@@ -3,27 +3,27 @@ on:
           echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
         shell: bash
-      - name: Restore module cache
+      - name: Cache Modules
```
s/Cache Modules/Restore Cached Go Modules? Maybe a better suggestion? But I think it should include "Go Modules"

shovnik replied:
I chose just cache over restore cache, because this not only restores the cache but also tells which path to save in cache after the job is complete. Renamed to "Cache Go Modules" as requested.

bogdandrutu commented on .github/workflows/build-and-test.yml:
```diff
@@ -96,24 +82,23 @@ jobs:
         with:
           go-version: 1.15
-      - name: Setup env
+      - name: Setup Environment
         run: |
           echo "GOPATH=$(go env GOPATH)" >> $GITHUB_ENV
           echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
```
s/Setup Environment/Setup Go Environment?

shovnik replied:
Renamed as requested.

bogdandrutu reviewed (commented):
There are lots of changes to already existing jobs (improvements in names, etc.) as well as adding new jobs. Consider to split this into 2 PRs, 1 for cosmetic/updates to existing jobs, and 1 for adding new jobs.

shovnik commented:
Rebased and extracted changes to existing jobs into separate PR that needs to be merged first as requested: https://github.com/open-telemetry/opentelemetry-collector/pull/2313. @bogdandrutu

bogdandrutu commented:
Please rebase 👯

shovnik commented:
Rebased

bogdandrutu commented:
> When merging this PR, once the repository is ready to move publishing from CircleCI to GHA, a maintainer will need to add the DOCKER_USERNAME and DOCKER_PASSWORD secrets for the opentelemetry docker account to allow pushing to docker. In addition a token with all access writes will need to be generated and added as a GITHUB_TOKEN secret to allow releases.

Please create an issue for this or document in the main issue where we track this.

shovnik commented:
I just added the comment to the main issue. A subsequent PR removing CircleCI workflows from the repository will also have to be made once we have validated that the GHA workflows replace them entirely. We will also make a PR tomorrow migrating the Collector-contrib workflows to GHA (more or less identical to core repo workflow). The main issue will not be closed until both the contrib repo has been migrated and the CircleCI workflows have been removed so everything should be remain tracked.
